torch/fx/experimental/graph_gradual_typechecker.py

```diff
@@ -0,0 +1,114 @@
+from functools import reduce
+import torch
+from torch.fx.tensor_type import Dyn, is_consistent, TensorType, is_more_precise
+import operator
+
+
+class GraphTypeChecker:
+    def __init__(self, env, traced):
+        self.env = env
+        self.traced = traced
+
+    def type_check(self):
+        """
+        A gradual type checker for graphs
+        Effect: every node's field type will be
+        populated with a type after type-checking is done
+        """
+        # here we collect types for variables
+        graph = self.traced.graph
+
+        # type check every node with gradual type rules
+        # if any node does not type check return false
+        for n in graph.nodes:
+            if not self.type_check_node(n):
+                return False
+        return True
+
+    def type_check_node(self, n):
+        """
+        Type check a given node.
+        Current operations:
+        - Reshape
+        - Transpose
+        - Add
+        """
+        if n.op == 'placeholder':
+            if n.type is None:
+                n.type = Dyn
+            else:
+                return n.type
+
+        if n.op == 'call_function':
+            if n.target == torch.add or n.target == operator.add:
+
+                t1 = n.args[0].type
+                t2 = n.args[1].type
+
+                if is_consistent(t1, t2):
+                    # we return the more precise type
+                    if is_more_precise(t1, t2):
```

Comment:
So is this what you were referring to in the meeting about putting the more precise type? I was thinking of that in terms of propagating back, but I think propagating forward, you would actually need to encode semantics for broadcasting. For example, you could have shape `1` on one side but `Dyn` on the other, but depending on the situation the `1` could be broadcasted out to the dynamic size

Reply:
broadcasting is done, let me know if the current definitions and tests are okay!